api/handlers.go

```diff
@@ -165,3 +171,404 @@ func (api *RecipeAPI) AddRecipeHandler(w http.ResponseWriter, req *http.Request)
+	defer req.Body.Close()
+	if err != nil {
+		log.Event(ctx, "error in reading request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Unmarshal to the shape of Instance struct
+	var instance recipe.Instance
+	err = json.Unmarshal(b, &instance)
+	if err != nil {
+		log.Event(ctx, "error returned from json unmarshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	//Validation to check if all the instance fields are entered
+	err = instance.ValidateAddInstance()
+	if err != nil {
+		log.Event(ctx, "bad request error as invalid instance given in request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Get currentRecipe from ID given to get instance of the recipe
+	currentRecipe, err := api.dataStore.Backend.GetRecipe(recipeID)
+	if err != nil {
+		log.Event(ctx, "error retrieving specific recipe from mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Append new instance into the current instance of the recipe
+	currentRecipe.OutputInstances = append(currentRecipe.OutputInstances, instance)
+
+	// Update the current recipe in mongo with the updated instance
+	update := bson.M{"$set": currentRecipe}
+	err = api.dataStore.Backend.UpdateRecipe(recipeID, update)
+	if err != nil {
+		log.Event(ctx, "error updating recipe to mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Marshal to output the newly added recipe
+	output, err := json.Marshal(currentRecipe)
+	if err != nil {
+		log.Event(ctx, "error returned from json marshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("content-type", "application/json")
+	w.Write(output)
+}
+
+//AddCodelistHandler - add a codelist in the instance of an existing recipe
+func (api *RecipeAPI) AddCodelistHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	//Get Update Recipe ID
+	vars := mux.Vars(req)
+	recipeID := vars["id"]
+	instanceID := vars["instance_id"]
+	logD := log.Data{"recipe_id": recipeID, "instance_id": instanceID}
+
+	// Read body
+	b, err := ioutil.ReadAll(req.Body)
+	defer req.Body.Close()
+	if err != nil {
+		log.Event(ctx, "error in reading request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Unmarshal to the shape of Codelist struct
+	var codelist recipe.CodeList
+	err = json.Unmarshal(b, &codelist)
+	if err != nil {
+		log.Event(ctx, "error returned from json unmarshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Generate the HRef of codelist if not given in request body as it follows a consistent pattern
+	hrefURL := "http://localhost:22400/code-lists/"
+	if codelist.HRef == "" {
+		codelist.HRef = hrefURL + instanceID
```

Comment:
a codelist href should never contain the instanceID, it would contain the codelist ID in the recipe.CodeList struct

Reply:
you may want to do this as part of or after the ValidateAddCodelist call, so youll be guaranteed to have an ID before trying to make the href

Reply:
Done in commit "update codelist.href validation and add tests to recipe package"